### PR TITLE
SNO+: Make HV/relay controls OWL aware

### PR DESCRIPTION
**Pull request by BenLand100**
Adds some class methods to ORXL3Model for storing the state of the
OWL supply (16B) and an instance method which identifies crates with
OWL tubes connected (3,13,18). The relay controls are disabled for
the whole crate for crates with OWL tubes whenever 16B is ON. This
is the safest option but may not be strictly necessary.

Also corrects a bug where 16B checked the HV interlocks for crate
16. Now 16B checks all crates identified as having OWL tubes
connected for a good HV interlock prior to turning on the supply.

### CODE REVIEW

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_LinkController.m [hunk 1]
```diff
@@ -791,11 +785,7 @@ - (void) hvStatusChanged:(NSNotification*)aNote
 
     BOOL lockedOrNotRunningMaintenance = [gSecurity runInProgressButNotType:eMaintenanceRunType orIsLocked:ORXL3Lock];
 
-    //todo: add exception for OWLs
-    //if ([hvPowerSupplyMatrix selectedColumn] == 0 && [model hvASwitch]) {
-    if ([model hvASwitch] || //A ON or
-        (([model crateNumber] == 3 || [model crateNumber] == 13 || [model crateNumber] == 18) && //OWL crate
-         ([owl_crate_master hvBSwitch]))) { //and crate 16 HV B ON
```
Comment (BenLand100) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_LinkController.m]: This was not safe as 16B never notified the other XL3s when it was switched on, so the states would not be updated until some later point in time, leaving relay controls enabled even with 16B on.

**Comment by BenLand100** [comment]
Fixes #145 (pending testing)

**Comment by tlatorre-uchicago** [comment]
I can test this on tonight's shift if you think it's ready to go. Let me know.

**Comment by BenLand100** [comment]
Yeah it's in a testable state. Also make sure the button states for HV work as expected when switching supplies on - I modified the notifications to only fire if the state changes.

**Comment by tlatorre-uchicago** [comment]
I just tried turning on the power supply for crates 16B and 15 and got the following error:

```
XL3 crate 16 NOT turning ON the HV power supply.
XL3 crate 15 NOT turning ON the HV power supply.
```

File: Source/Objects/Custom Hardware/SNO+/XL3/ORXL3Model.m [hunk 2]
```diff
@@ -3788,24 +3839,49 @@ - (void) setHVSwitch:(BOOL)aOn forPowerSupply:(unsigned char)sup
...
+            if (!interlockIsGood) NSLogColor([NSColor redColor],@"%@ HV interlock BAD\n",[[self xl3Link] crateName]);
+        }
+        @catch (NSException *exception) {
+            NSLogColor([NSColor redColor],@"%@ error in readHVInterlock\n",[[self xl3Link] crateName]);
+        }
     }
-
-    if (!interlockIsGood) {
-        NSLogColor([NSColor redColor],@"%@ HV interlock BAD\n",[[self xl3Link] crateName]);
+    if (interlockIsGood) {
```
Comment (BenLand100) [inline Source/Objects/Custom Hardware/SNO+/XL3/ORXL3Model.m]: Doh. That's why.

**Comment by BenLand100** [comment]
Should be fixed.

**Comment by tlatorre-uchicago** [comment]
We did a few tests and everything seemed to work:

- turned HV on for 16B
- verified that couldn't set relays for 3, 13, or 18
- turned HV off for 16B
- verified that we could set relays for 3, 13, and 18
- turned HV on for 16A
- verified that we could set relays for 3, 13, and 18

merged.

**Pull request closed without merging**